bib_rights/import_bib_rights_data.pl

```diff
@@ -0,0 +1,290 @@
+    verbose  => $verbose,
+    instance => $instance
+);
+
+$secs = 600 unless defined $secs;
+$crms->set('noop', 1) if $noop;
+
+my $report = '';
+my $catalogPath = '';
+my $fileToProcess;
+my $lastCount = 0;
+my $lastUpdate = 0;
+my $type = 'full';
+
+my $sql = 'SELECT COUNT(*) FROM systemvars WHERE name=?';
+if ($crms->SimpleSqlGet($sql, 'catalogUpdateInProgress')) {
+  exit(0);
+}
+
+if (scalar @ARGV && -e $ARGV[0]) {
+  $fileToProcess = $ARGV[0];
+  $report .= "Using $fileToProcess from command line\n";
+  $type = 'manual';
+} else {
+  my $lastFull = $crms->GetSystemVar('lastCatalogImport');
+  my $lastDate;
+  if (defined $lastFull && $lastFull =~ m/^zephir_full_(\d{8})_vufind.json.gz$/i) {
+    $lastDate = $1;
+  }
+  $lastCount = $crms->GetSystemVar('lastCatalogImportCount');
+  $lastUpdate = $crms->GetSystemVar('lastCatalogUpdate');
+  if (defined $lastUpdate && $lastUpdate =~ m/^zephir_upd_(\d{8}).json.gz$/i) {
+    $lastDate = $1 if $1 gt $lastDate;
+  }
+  $catalogPath = '/htapps/archive/catalog/';
+
+  opendir(DIR, "$catalogPath") or die "Can't open $catalogPath\n";
+  my @files = sort readdir(DIR);
+  closedir(DIR);
+
+  my $nextImportFile;
+  my $nextUpdateFile;
+  foreach my $file (@files)
+  {
+    my $date;
+    if ($file =~ m/^zephir_full_(\d{8})_vufind\.json\.gz$/i &&
+        (!defined $lastFull || $file gt $lastFull)) {
+      $date = $1;
+      $nextImportFile = $file;
+    }
+    if ($file =~ m/^zephir_upd_(\d{8})\.json\.gz$/i && !defined $nextUpdateFile) {
+      $date = $1;
+      if (defined $lastDate && $date gt $lastDate) {
+        $nextUpdateFile = $file;
+      }
+    }
+  }
+  $report .= sprintf "Next import %s, next update %s\n",
+                     (defined $nextImportFile)? $nextImportFile:'[undef]',
+                     (defined $nextUpdateFile)? $nextUpdateFile:'[undef]';
+
+  # If there is a newer full file, abandon last file and start over with this one.
+  # Otherwise, if the most recent full dump is incomplete (systemvars.lastCatalogImportCount
+  # is present) then continue to work on that.
+  # Otherwise ingest the oldest update file after systemvars.lastCatalogUpdate
+  # (if it exists) without a timeout.
+  if (defined $nextImportFile) {
+    $report .=  "Moving on to next full import $nextImportFile<br/>\n";
+    $fileToProcess = $nextImportFile;
+    $lastCount = 0;
+  }
+  elsif (defined $lastCount) {
+    $report .=  "Continuing $lastFull at $lastCount<br/>\n";
+    $fileToProcess = $lastFull;
+  }
+  elsif (defined $nextUpdateFile) {
+    $report .= "Beginning update $nextUpdateFile<br/>\n";
+    $fileToProcess = $nextUpdateFile;
+    $type = 'update';
+  }
+}
+
+my $alarmFired = 0;
+local $SIG{ALRM} = sub { $report .= "ALARM FIRED<br/>\n"; $alarmFired = 1; };
+my $json_xs = JSON::XS->new;
+# De-spam the output of this script.
+exit(0) unless defined $fileToProcess;
+
+$sql = 'REPLACE INTO systemvars (name,value) VALUES (?,1)';
+$crms->PrepareSubmitSql($sql, 'catalogUpdateInProgress');
```

Comment:
I think I have a more up to date version of this running currently and IIRC I did some signal handling for just this reason. That requires a manual scp and update because of the way I have it deployed, alas. I'll grab the current version and commit it.

Reply:
Updated with signal handling as well as some minor edits. Not quite ready to add test fixtures for it, however ;-)